contracts/contracts/rem.system/src/producer_pay.cpp

```diff
@@ -182,30 +182,70 @@ namespace eosiosystem {
    }
 
    using namespace eosio;
-   void system_contract::claimrewards( const name& owner ) {
-      require_auth( owner );
-
-      const auto& prod = _producers.get( owner.value );
-      check( prod.active(), "producer does not have an active key" );
 
-      check( _gstate.total_activated_stake >= min_activated_stake,
-                    "cannot claim rewards until the chain is activated (at least 15% of all tokens participate in voting)" );
+   void system_contract::claim_perstake( const name& guardian )
+   {
+      const auto& voter = _voters.get( guardian.value );
 
       const auto ct = current_time_point();
+      check( ct - voter.last_claim_time > microseconds(useconds_per_day), "already claimed rewards within past day" );
+
+      if (_gstate.perstake_bucket > 1'0000) { //do not distribute small amounts
+         int64_t total_per_stake_pay = 0;
+
+         auto sorted_voters = _voters.get_index<"bystake"_n>();
+         _gstate.total_guardians_stake = 0;
+         for (auto it = sorted_voters.rbegin(); it != sorted_voters.rend() && it->staked >= _gremstate.producer_stake_threshold; it++) {
+            if ( vote_is_reasserted( it->last_reassertion_time ) ) {
+               _gstate.total_guardians_stake += it->staked;
+            }
+         }
+
+         for (auto it = sorted_voters.rbegin(); it != sorted_voters.rend() && it->staked >= _gremstate.producer_stake_threshold; it++) {
+            if ( vote_is_reasserted( it->last_reassertion_time ) ) {
+               const int64_t per_stake_pay = _gstate.perstake_bucket * (double(it->staked) / double(_gstate.total_guardians_stake));
+               
+               _voters.modify( *it, same_payer, [&](auto &v) {
+                  v.pending_perstake_reward += per_stake_pay;
+               });
+               
+               total_per_stake_pay += per_stake_pay;
+            }
+         }
+
+         _gstate.perstake_bucket -= total_per_stake_pay;
+         check(_gstate.perstake_bucket >= 0, "perstake_bucket cannot be negative");
+      }
+      const auto per_stake_pay = voter.pending_perstake_reward;
+
+      _voters.modify( voter, same_payer, [&](auto& v) {
+         v.last_claim_time         = ct;
+         v.pending_perstake_reward = 0;
+      }); 
+
+      if ( per_stake_pay > 0 ) {
+         token::transfer_action transfer_act{ token_account, { {spay_account, active_permission}, {guardian, active_permission} } };
+         transfer_act.send( spay_account, guardian, asset(per_stake_pay, core_symbol()), "producer stake pay" );
+      }
+   }
 
+   void system_contract::claim_pervote( const name& producer )
+   {
+      const auto& prod = _producers.get( producer.value );
+      check( prod.active(), "producer does not have an active key" );
```

Comment:
I think inactive BPs also should be able to claim rewards they earned in the past.

Reply:
this is check from EOS, I can remove it but not sure what will be affected. As for me it's not a big deal for producer to check for pending rewards before deactivation.

Reply:
will Guardian be able to claim his reward if he used to be a BP in the past but is not a BP any longer?